hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/MembershipNamenodeResolver.java

```diff
@@ -478,4 +476,27 @@ private List<MembershipState> getRecentRegistrationForQuery(
   public void setRouterId(String router) {
     this.routerId = router;
   }
+
+  /**
+   * Shuffle cache, to ensure that the current nn will not be accessed first next time.
+   *
+   *
+   * @param nsId name service id
+   * @param namenode namenode contexts
+   */
+  @Override
+  public synchronized void shuffleCache(String nsId, FederationNamenodeContext namenode) {
+    cacheNS.compute(Pair.of(nsId, false), (ns, namenodeContexts) -> {
+      if (namenodeContexts != null
+              && namenodeContexts.size() > 0
+              && !namenodeContexts.get(0).getState().equals(ACTIVE)
+              && namenodeContexts.get(0).getRpcAddress().equals(namenode.getRpcAddress())) {
+        List<FederationNamenodeContext> rotatedNnContexts = new ArrayList<>(namenodeContexts);
+        Collections.rotate(rotatedNnContexts, -1);
```

Comment:
We are not really shuffling right?

Reply:
The purpose of this method is to reduce the priority of the next access to the wrong nn in this request. In the default implementation of MembershipNamenodeResolver, we can directly rotate the cache to the left, because the cache of ns in the MembershipNamenodeResolver is a list, but considering the implementation of other NamenodeResolver (Maybe there will be in the future), so this name may be changed to a more intuitive method name.

Reply:
rotateCache looks better.